# LAB | Feature Engineering

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In [28]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

**Check the shape of your data**

In [3]:
#your code here
spaceship.shape

(8693, 14)

**Check for data types**

In [4]:
#your code here
spaceship.dtypes

PassengerId      object
HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Name             object
Transported        bool
dtype: object

**Check for missing values**

In [5]:
#your code here
spaceship.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB


There are multiple strategies to handle missing data

- Removing all rows or all columns containing missing data.
- Filling all missing values with a value (mean in continouos or mode in categorical for example).
- Filling all missing values with an algorithm.

For this exercise, because we have such low amount of null values, we will drop rows containing any missing value. 

In [6]:
#your code here
spaceship.dropna(inplace=True)

- **Cabin** is too granular - transform it in order to obtain {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T'}

In [9]:
#your code here
spaceship['Cabin'] = spaceship['Cabin'].apply(lambda x: str(x)[0] if pd.notna(x) else x)
spaceship['Cabin'].value_counts()

Cabin
F    2152
G    1973
E     683
B     628
C     587
D     374
A     207
T       2
Name: count, dtype: int64

- Drop PassengerId and Name

In [10]:
#your code here
spaceship = spaceship.drop(columns = ["PassengerId", "Name"])
spaceship

HomePlanet CryoSleep Cabin    Destination   Age    VIP  RoomService  \
0        Europa     False     B    TRAPPIST-1e  39.0  False          0.0   
1         Earth     False     F    TRAPPIST-1e  24.0  False        109.0   
2        Europa     False     A    TRAPPIST-1e  58.0   True         43.0   
3        Europa     False     A    TRAPPIST-1e  33.0  False          0.0   
4         Earth     False     F    TRAPPIST-1e  16.0  False        303.0   
...         ...       ...   ...            ...   ...    ...          ...   
8688     Europa     False     A    55 Cancri e  41.0   True          0.0   
8689      Earth      True     G  PSO J318.5-22  18.0  False          0.0   
8690      Earth     False     G    TRAPPIST-1e  26.0  False          0.0   
8691     Europa     False     E    55 Cancri e  32.0  False          0.0   
8692     Europa     False     E    TRAPPIST-1e  44.0  False        126.0   

      FoodCourt  ShoppingMall     Spa  VRDeck  Transported  
0           0.0           0.0     0.0     0.0        False  
1           9.0          25.0   549.0    44.0         True  
2        3576.0           0.0  6715.0    49.0        False  
3        1283.0         371.0  3329.0   193.0        False  
4          70.0         151.0   565.0     2.0         True  
...         ...           ...     ...     ...          ...  
8688     6819.0           0.0  1643.0    74.0        False  
8689        0.0           0.0     0.0     0.0        False  
8690        0.0        1872.0     1.0     0.0         True  
8691     1049.0           0.0   353.0  3235.0        False  
8692     4688.0           0.0     0.0    12.0         True  

[6606 rows x 12 columns]

- For non-numerical columns, do dummies.

In [11]:
#your code here
spaceship = pd.get_dummies(spaceship, columns=['HomePlanet', 'Cabin', 'Destination'])

In [24]:
spaceship

CryoSleep   Age    VIP  RoomService  FoodCourt  ShoppingMall     Spa  \
0        False  39.0  False          0.0        0.0           0.0     0.0   
1        False  24.0  False        109.0        9.0          25.0   549.0   
2        False  58.0   True         43.0     3576.0           0.0  6715.0   
3        False  33.0  False          0.0     1283.0         371.0  3329.0   
4        False  16.0  False        303.0       70.0         151.0   565.0   
...        ...   ...    ...          ...        ...           ...     ...   
8688     False  41.0   True          0.0     6819.0           0.0  1643.0   
8689      True  18.0  False          0.0        0.0           0.0     0.0   
8690     False  26.0  False          0.0        0.0        1872.0     1.0   
8691     False  32.0  False          0.0     1049.0           0.0   353.0   
8692     False  44.0  False        126.0     4688.0           0.0     0.0   

      VRDeck  Transported  HomePlanet_Earth  ...  Cabin_B  Cabin_C  Cabin_D  \
0        0.0        False             False  ...     True    False    False   
1       44.0         True              True  ...    False    False    False   
2       49.0        False             False  ...    False    False    False   
3      193.0        False             False  ...    False    False    False   
4        2.0         True              True  ...    False    False    False   
...      ...          ...               ...  ...      ...      ...      ...   
8688    74.0        False             False  ...    False    False    False   
8689     0.0        False              True  ...    False    False    False   
8690     0.0         True              True  ...    False    False    False   
8691  3235.0        False             False  ...    False    False    False   
8692    12.0         True             False  ...    False    False    False   

      Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0       False    False    False    False                    False   
1       False     True    False    False                    False   
2       False    False    False    False                    False   
3       False    False    False    False                    False   
4       False     True    False    False                    False   
...       ...      ...      ...      ...                      ...   
8688    False    False    False    False                     True   
8689    False    False     True    False                    False   
8690    False    False     True    False                    False   
8691     True    False    False    False                     True   
8692     True    False    False    False                    False   

      Destination_PSO J318.5-22  Destination_TRAPPIST-1e  
0                         False                     True  
1                         False                     True  
2                         False                     True  
3                         False                     True  
4                         False                     True  
...                         ...                      ...  
8688                      False                    False  
8689                       True                    False  
8690                      False                     True  
8691                      False                    False  
8692                      False                     True  

[6606 rows x 23 columns]

**Perform Train Test Split**

In [25]:
#your code here
features = spaceship.drop(columns = ['Transported'])
target = spaceship["Transported"]

In [26]:
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size = 0.20, random_state=22)

In [27]:
print('X_train shape is:',X_train.shape)
print('y_train shape is:', y_train.shape)
print('X_test shape is:', X_test.shape)
print('y_train shape is:', y_test.shape)

X_train shape is: (5284, 22)
y_train shape is: (5284,)
X_test shape is: (1322, 22)
y_train shape is: (1322,)


**Model Selection**

In this exercise we will be using **KNN** as our predictive model.

In [29]:
#your code here
knn = KNeighborsClassifier()


- Evaluate your model's performance. Comment it

In [30]:
#your code here
knn.fit(X_train, y_train)

KNeighborsClassifier()

In [31]:
knn.score(X_test, y_test)

0.7934947049924357

In [33]:
pred = knn.predict(X_test)
pred

array([ True,  True, False, ...,  True,  True, False])

In [34]:
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score

In [35]:
print(f"Precision:",{precision_score(y_test, pred,zero_division=0)})
print(f"Accuracy:",{accuracy_score(y_test, pred)})
print(f"Recall:",{recall_score(y_test, pred)})
print(f"F1:",{f1_score(y_test, pred)})

Precision: {0.8087349397590361}
Accuracy: {0.7934947049924357}
Recall: {0.7862371888726208}
F1: {0.7973273942093542}


Even though dropping features improved the model performance of the lecture model example, it seems that for this model, dropping features improved the score from 0.76 to 0.79.
Additionally, the performance metrics are quite good with a precision of 80%, a recall score of 78%, and an F1 score of 79%.